### PR TITLE
Add **kwargs support to Bag.map and Bag.map_partitions

### PR DESCRIPTION
**Pull request by jcorbin**
Allows to pass non-dask python literals and `dask.bag.Item`s as values in **kwargs which pass through to the map func.

### CODE REVIEW

File: dask/bag/core.py [hunk 1]
```diff
@@ -183,6 +183,42 @@ def finalize_item(results):
...
+    return dsk, kw_pairs
+
+
+def with_kwargs(func, kwargs):
+    """ Wraps func so that it gets called with **kwargs
+
+    Useful for deferring a partial(func, **kwargs) with eventually-computed
+    kwargs in a dask graph.
+
+    TODO: if core supported splicing args for us, we could replace usage of
```
Comment (jcorbin) [inline dask/bag/core.py]: Feedback welcome here, this smelled like "too much technology" similarly to #1044 and so I wrote it down as scope creep for this PR.  Either way (be it computed functions or arg splicing) I put it to you that you're missing a crucial part in your "incomplete lisp" around this pain point ;-)

File: dask/bag/core.py [hunk 2]
```diff
@@ -183,6 +183,42 @@ def finalize_item(results):
...
+    a dask graph to pass the final computed **kwargs to a function.
+
+    Currently only dask.bag.Item and python literal values are supported.
+    """
+    kw_pairs = []
+    for key, val in iteritems(kwargs):
+        if isinstance(val, Item):
+            dsk = merge(dsk, val.dask)
+            val = val.key
+        # TODO elif isinstance(val, Value):
```
Comment (jcorbin) [inline dask/bag/core.py]: I've intentionally left this (and e.g. dask.dataframe.Scalar) out for now simply because I've had enough surprises of my own when crossing scheduler boundaries to be wary of just allowing this by default.  However there is quite a bit of missing sugar for users to tell dask "no really please do it" explicitly; things like:
- `bag.item.to_imperative`
- `imperative.value.to_item`
- `dataframe.scalar.to_` anything

Despite what regret there may be about having multiple types for "a scalar computation value", I think that we should at least add commuting methods between all three (that I've found so far at least) competing types for such a thing ;-)

File: dask/bag/core.py [hunk 3]
```diff
@@ -297,19 +352,37 @@ def remove(self, predicate):
...
+        1.0
         """
         name = 'map-partitions-{0}-{1}'.format(funcname(func),
-                                               tokenize(self, func))
-        dsk = dict(((name, i), (func, (self.name, i)))
+                                               tokenize(self, func, kwargs))
+        dsk = self.dask.copy()
+        dsk, kw_pairs = merge_kwargs(dsk, kwargs)
+        dsk.update(((name, i),
+                    (apply, func, [(self.name, i)], (dict, kw_pairs)))
```
Comment (jcorbin) [inline dask/bag/core.py]: open question of should I just use `func = (with_kwargs, ...)` here too; I wrote the .map code and its necessary `with_kwargs` utility last and this one first, so I didn't use it here; feedback welcome :-)

File: dask/bag/tests/test_bag.py [hunk 4]
```diff
@@ -300,6 +306,42 @@ def test_map_partitions():
     assert b.map_partitions(lambda a: len(a) + 1).name != b.map_partitions(len).name
 
 
+def test_map_partitions_with_kwargs():
+    corpus = '''
```
Comment (jcorbin) [inline dask/bag/tests/test_bag.py]: I rather like this test case since it's much closer to my original use case, and not a contrived number normalizer; however let me know if it's too weird to keep...

**Comment by jcorbin** [comment]
Supercedes #1041

File: dask/bag/core.py (hunk 3 above)
Comment (mrocklin) [inline dask/bag/core.py]: Would using just `apply` work here instead?  Generally it's nice to keep information in the dask graph rather than hidden away in `partial`s or closures.  This makes debugging much easier.  It is also one less function that dask developers need to keep track of in the future.

File: dask/bag/core.py [hunk 5]
```diff
@@ -183,6 +183,42 @@ def finalize_item(results):
...
+        if isinstance(val, Item):
+            dsk = merge(dsk, val.dask)
+            val = val.key
+        # TODO elif isinstance(val, Value):
+        elif isinstance(val, Base):
+            raise NotImplementedError(
+                '%s not supported as kwarg value to Bag.map_partitions'
+                % type(val).__name__)
+        kw_pairs.append([key, val])
+    return dsk, kw_pairs
```
Comment (mrocklin) [inline dask/bag/core.py]: This function strikes me as very special purpose.  I can't imagine how it would be useful outside of bag.map.  I wonder if, instead, we could generalize and reuse `to_task_dasks` to serve our purpose here.

File: dask/bag/core.py [hunk 6]
```diff
@@ -244,20 +280,39 @@ def __init__(self, dsk, name, npartitions):
...
+
+        If a dask.Bag.Item is used its computed value is passed to the map
+        function.  This means that whatever computations are implied by any
+        Item values block any of the map work.  Example:
+
+        >>> b = db.from_sequence(range(5))
+        >>> b.map(
+        ...     lambda x, total=0: float(x) / total,
+        ...     total=b.sum()).sum().compute()
+        1.0
```
Comment (mrocklin) [inline dask/bag/core.py]: This example takes me a moment to grok and so, while it's a fine test, it's a bit odd for documentation.  I wonder if an example like the following might be easier on the reader.

```python
>>> def add(x=0, y=0):
...     return x + y

>>> b = db.from_sequence(range(5))
>>> b.map(add, y=10).compute()
(10, 11, 12, 13, 14)
```

File: dask/bag/tests/test_bag.py [hunk 7]
```diff
@@ -300,6 +306,42 @@ def test_map_partitions():
...
+        perpartition=lambda part: set(part),
+        aggregate=lambda sets: reduce(set.union, sets))
+    assert all_the_words.compute() == set(words)
+
+    coded_word_bag = word_bag.map_partitions(
+        code_words, all_words=all_the_words)
+
+    expected_code = dict((word, code)
+                         for code, word in enumerate(sorted(set(words))))
+    print(expected_code)
```
Comment (mrocklin) [inline dask/bag/tests/test_bag.py]: left over print statement

File: dask/bag/tests/test_bag.py [hunk 8]
```diff
@@ -300,6 +306,42 @@ def test_map_partitions():
...
+    assert coded_word_bag.compute(get=dask.get) == [expected_code[word] for word in words]
```
Comment (mrocklin) [inline dask/bag/tests/test_bag.py]: This unittest seems like it has some complexity that is unrelated to the code.  A future developer might find that they have to spend a minute or two understanding your application in order to properly debug a failing test in the future.  Is there a simpler set of examples we can test instead?  I find basic arithmetic to be fairly broadly accessible.

File: dask/bag/core.py (hunk 5 above)
Comment (jcorbin) [inline dask/bag/core.py]: Aye, shall I salvage as mentioned in https://github.com/dask/dask/pull/1041#issuecomment-197038147 first?  I was perhaps over recalibrating here since I am indeed prone to over solving things ;-)

Comment (mrocklin) [inline dask/bag/core.py]: Maybe?  What is the smallest possible change we can make to to_task_dasks
to achieve the desired effect?

File: dask/bag/core.py (hunk 6 above)
Comment (jcorbin) [inline dask/bag/core.py]: Sure I can totally see having an example in doc string with a constant python value, but we also want one with a dask value as well, I think simply using temporary variables to make the code clearer, rather than going along with the penchant for heavy inlining that dask's source code has would help:
```python
>>> nums = db.from_sequence(...)
>>> nums_total = nums.sum()
>>> normalized = b.map(lambda x, total=0: float(x) / total, total=nums_total)
>>> normalized.sum()
1.0
```

File: dask/bag/core.py (hunk 3 above)
Comment (jcorbin) [inline dask/bag/core.py]: wait this one is the one that uses apply, I think you meant to comment on the `Bag.map` diff hunk?

File: dask/bag/core.py (hunk 6 above)
Comment (mrocklin) [inline dask/bag/core.py]: I think we should start with the literal value.  This gets the common case
out first, which I think is as if not more important.

Adding something like b.sum() sounds fine too.  I recommend keeping it
inline but moving the lambda out, perhaps replace it with a function div?
I wouldn't worry about `float(x)`, from __future__ import division should
have you covered.

File: dask/bag/core.py [hunk 9]
```diff
@@ -244,20 +280,39 @@ def __init__(self, dsk, name, npartitions):
...
-        name = 'map-{0}-{1}'.format(funcname(func), tokenize(self, func))
+        name = 'map-{0}-{1}'.format(funcname(func),
+                                    tokenize(self, func, kwargs))
         if takes_multiple_arguments(func):
             func = partial(apply, func)
-        dsk = dict(((name, i), (reify, (map, func, (self.name, i))))
+        dsk = self.dask.copy()
+        if kwargs:
+            dsk, kw_pairs = merge_kwargs(dsk, kwargs)
+            func = (with_kwargs, func, (dict, kw_pairs))
```
Comment (jcorbin) [inline dask/bag/core.py]: Okay, so here we are with the place where "can we just use apply" applies I think, my thoughts:
- I wrote `with_kwargs` explicitly so it gets called in the graph, rather than proliferate on the existing func obfuscation that the multi-args branch uses
- the only other alternative that I can think of would be a more specialized `map_with_kwargs`
- the underlying problem here is that we have to unwrap the partition to match func's type signature...
- ...I originally chosen to not modify the existing method-of-unwrapping, and instead to do an eventual function wrapping, only possible here because func itself is not being directly called, obviating the need for computable function technology in core ;-)

However I could see that it would be perhaps clearer to branch the graph to either be one of:
```python
# in the old case
(map, func, (self.name, i))

# in the new case (arg-ordering bikesheds welcome):
(map_with_kwargs, func, (dict, kw_pairs), (self.name, i))
```

Comment (mrocklin) [inline dask/bag/core.py]: Ah, I see, so `(with_kwargs, func, kwargs)` is the same as `(apply, partial, (func,), kwargs)`?

File: dask/bag/core.py (hunk 5 above)
Comment (jcorbin) [inline dask/bag/core.py]: I'm not certain that we actually want the full machinery of `to_task_dasks` here, unless we really do want people to be able to pass fully-realized bags, dataframes, arrays, etc along... which seems like a good way to get horrible performance presuming you don't blow out memory ;-)

But let's say we do:
-`to_task_dasks` is from the imperative side at present, it only treats `Value`s natively (everything else is optimize/inlined) which causes expensive computations (e.g. shard loading) to get duplicated in my realistic two-pass use-case
- this was why in https://github.com/dask/dask/pull/1041/commits I first factored a generalized core out of the imperative version, and then introduced a bag-native version reusing the same python-object-walking core

Maybe what's needed is more clarity around the `to_task_dasks` work started in #1041.  For example rather than the weird, but Works For Me "special" support of the [new `base.normalize_to_dasks` core](https://github.com/dask/dask/pull/1041/commits/9a818bfdcd306659c29dd18ac6bb93afb008c715):
  - it should have a way to differentiate between which dask types get referenced by key (error if a referenceable type has more than one key) vs optimized and inlined
  - it could have a way to blacklist certain types (presuming there's no better way to say "all large dasks" if that's a thing we want)

File: dask/bag/core.py (hunk 9 above)
Comment (jcorbin) [inline dask/bag/core.py]: Hmm, it may well be, I just didn't trust it up against the existing partial-fuscation of func in context, I'll see if that works locally :-)

**Comment by jcorbin** [comment]
Updated the docstrings, will squash before merge.

**Comment by jcorbin** [comment]
Okay, turns out that `(with_kwargs, func, kwargs)` is equivalent to `(apply, partial, [func], kwargs)`.

File: dask/bag/core.py (hunk 5 above)
Comment (mrocklin) [inline dask/bag/core.py]: OK, that makes sense.  Thoughts on the following?

```python
def unpack_kwargs_for_map(kwargs):
    dsk = {}
    for k, v in kwargs.items():
        if hasattr(v, 'dask'):
            if hasattr(v, 'key'):
                kwargs[k] = v.key
                dsk.update(v.dask)
            else:
                raise ValueError(...)
    return dsk, kwargs
```
    
This has the benefit of supporting both `Value`s and `Item`s and it's also pure and so easily testable.

Unfortunately, this still lacks support for any sort of nesting, so if someone does the following then they'll get unfriendly behavior.

    b.map(func, keyword=[b.sum()])

This nesting case is where something more general could really be handy.  I agree with you though, that the fact that we want to limit ourselves by the contained dask collections makes things trickier.

I wonder if @jcrist has a suggestion.

Comment (jcorbin) [inline dask/bag/core.py]: Sure, returning a new dict that the caller could then merge is an easy change.... what's the tradeoff between putting a dict-literal in the graph and a `(dict, [[k, v]...])` which is how `to_task_dasks` represents eventual-dicts?

**Comment by jcorbin** [comment]
Okay, reworked merge_kwargs => unpack_kwargs (#willsquashbeforemerge); I didn't name it with the suggested `_for_map` suffix since there's nothing map-specific about it, and a subsequent diff could extend reduction to allow passing kwargs for perpartition (however I'm intentionally keeping this one focused on the map functions).

File: dask/bag/core.py (hunk 5 above)
Comment (jcorbin) [inline dask/bag/core.py]: I'm dubious about loosening the referencing mechanism to be duck-typed because of the schedule-boundary-crossing problem.  We can always loosen it later, and/or reconcile this function and to_task_dasks going forward.

Comment (jcorbin) [inline dask/bag/core.py]: Also note that the duck typing isn't even sufficient for all "natural" values currently; e.g. `dataframe.Scalar` lacks a key property...

Comment (mrocklin) [inline dask/bag/core.py]: OK, makes sense.  Can I put you on the hook to handle this if a user ever asks about why this doesn't work with `Value` objects?

Regarding `dict` it looks like that was a mistake on my part.  I was assuming that the dask scheduler walked down dict objects.  Looks like it doesn't at the moment.

```python
In [1]: import dask

In [2]: def inc(x):
    return x + 1
   ...: 

In [3]: dsk = {'x': 1, 'y': {'a': 'x'}}

In [4]: dask.get(dsk, 'y')
Out[4]: {'a': 'x'}
```

Comment (jcorbin) [inline dask/bag/core.py]: Sure thing, count me on hook :-)

**Comment by jcorbin** [comment]
Okay rebased and squashed on latest master.

File: dask/bag/tests/test_bag.py (hunk 4 above)
Comment (mrocklin) [inline dask/bag/tests/test_bag.py]: If its cheap to run then I won't say no to a test.  However, we do need a simpler unit test that is easy for the maintainer to quickly understand at 4am when things break and need to be resolved quickly.

File: dask/bag/tests/test_bag.py [hunk 10]
```diff
@@ -103,6 +103,12 @@ def test_map_with_builtins():
         [True, False]
 
 
+def test_map_with_kwargs():
+    b = db.from_sequence(range(100), npartitions=10)
+    assert b.map(lambda x, total=0: float(x) / total,
+                 total=b.sum()).sum().compute() == 1.0
```
Comment (mrocklin) [inline dask/bag/tests/test_bag.py]: We should test with literals as well

**Comment by mrocklin** [comment]
This is all looking pretty good to me.  I left some final comments on testing.

**Comment by jcorbin** [comment]
Updated tests:
- added literal tests for Bag.map
- shifted tests for Bag.map_partitions to be direct lifts of the Bag.map cases

Will squash before final merge mudolo feedback.

**Comment by mrocklin** [comment]
+1

@jcrist any comments?

File: dask/bag/core.py [hunk 11]
```diff
@@ -183,6 +183,29 @@ def finalize_item(results):
...
+    Currently only dask.bag.Item and python literal values are supported.
+
+    Returns a merged dask graph and a list of [key, val] pairs suitable for
+    eventually constructing a dict.
+    """
+    dsk = {}
+    kw_pairs = []
+    for key, val in iteritems(kwargs):
+        if isinstance(val, Item):
+            dsk = merge(dsk, val.dask)
```
Comment (jcrist) [inline dask/bag/core.py]: Nitpick, could actually be `dsk.update(val.dask)` - `merge` creates a new dict and copies the old values over. Slightly more efficient, but not super important.

File: dask/bag/core.py [hunk 12]
```diff
@@ -244,20 +267,53 @@ def __init__(self, dsk, name, npartitions):
...
-        name = 'map-{0}-{1}'.format(funcname(func), tokenize(self, func))
+        name = 'map-{0}-{1}'.format(funcname(func),
+                                    tokenize(self, func, kwargs))
         if takes_multiple_arguments(func):
             func = partial(apply, func)
-        dsk = dict(((name, i), (reify, (map, func, (self.name, i))))
+        dsk = self.dask.copy()
+        if kwargs:
+            kw_dsk, kw_pairs = unpack_kwargs(kwargs)
+            dsk = merge(dsk, kw_dsk)
```
Comment (jcrist) [inline dask/bag/core.py]: Same here. You can just mutate `dsk` instead of creating a new one. Not super important, as time difference will be negligible.

File: dask/bag/core.py (hunk 11 above)
Comment (mrocklin) [inline dask/bag/core.py]: I was groaning about impure use earlier, which is possibly why @jcorbin is doing it this way.  I agree with @jcrist though that, because we're only mutating a locally created `dict`, rather than something that was passed in, `update` is probably slightly preferred, especially as graphs can become large.

File: dask/bag/core.py [hunk 13]
```diff
@@ -244,20 +267,53 @@ def __init__(self, dsk, name, npartitions):
...
+        Item values block any of the map work.
+
+        For example the above ended up doing its two passes over the data in
+        two separate dask computations.  If the data were expensive to access
+        (rather than being an in-memory range) we would instead want to combine
+        both passes over the data into one shared computation:
+
+        >>> normalized = b.map(normalize, total=b.sum())
+        >>> normalized.sum().compute()
+        1.0
```
Comment (jcrist) [inline dask/bag/core.py]: This docstring took me a bit to fully grok, which means it could probably be cleaner. Perhaps:

```python
"""
...

Keyword arguments are passed through to `func`. These can be either 
`dask.bag.Item`s, or normal python objects.

Example
-------
>>> b = db.from_sequence(range(1, 101), npartitions=10)
>>> def div(num, den=1):
...     return num/den

Using a python object:

>>> den = b.max().compute()
>>> den
100
>>> b.map(div, den=den).take(5)
[0.01, 0.02, 0.03, 0.04, 0.05]

Using an `Item`:

>>> b.map(div, den=b.max()).take(5)
[0.01, 0.02, 0.03, 0.04, 0.05]

Note that while both versions give the same output, the second forms 
a single graph, and then computes everything at once, and in some 
cases may be more efficient.
"""
```

**Comment by jcrist** [comment]
Sorry for coming in late and complaining after you've fixed things :). I've left some comments, but none of them are 100% blocking.

File: dask/bag/core.py (hunk 11 above)
Comment (jcorbin) [inline dask/bag/core.py]: will change, I was only using merge out of prevailing idiom :-)

**Comment by jcorbin** [comment]
- updated to use dict.update rather than merge where noted
- new docstring based around numeric max / take(5)

Will squash pending review once we're happy with the new docstrings

**Comment by jcrist** [comment]
This looks good (and useful) to me. Thanks for making those changes. +1, will merged once squashed.

**Comment by jcorbin** [comment]
Squashed, merge at will.

**Comment by mrocklin** [comment]
I'm very glad to see this in.  Thanks @jcorbin !
